### PR TITLE
Attempts to fix issue #266

### PR DESCRIPTION
**Pull request by keyehzy**
We do this by treating `else (` separately. If `else` is followed by
a (conditional + block) pattern, maybe the intention was to use an
`else if` instead and a warning is emitted.

### CODE REVIEW

File: src/quick-lint-js/error.h [hunk 1]
```diff
@@ -214,6 +214,13 @@
       .error(QLJS_TRANSLATABLE("'else' has no corresponding 'if'"),            \
              else_token))                                                      \
                                                                                \
+  QLJS_ERROR_TYPE(                                                             \
+      error_else_with_conditional_missing_if, "E???",                          \
```
Comment (strager) [inline src/quick-lint-js/error.h:218]: Blocking: Pick a valid error code. The `???` should be three decimal digits. (I should write a test to ensure error codes are well-formed.)

File: test/test-parse-statement.cpp [hunk 2]
```diff
@@ -578,6 +578,42 @@ TEST(test_parse, else_without_if) {
...
+                                      "visit_variable_use",        // d
+                                      "visit_exit_block_scope"));  // </block>
+
+    EXPECT_THAT(
+        v.errors,
+        ElementsAre(ERROR_TYPE_FIELD(
+            error_else_with_conditional_missing_if, else_token,
+            offsets_matcher(&code, strlen(u8"if (a) { b; } "), u8"else"))));
+  }
+}
```
Comment (strager) [inline test/test-parse-statement.cpp:615]: Blocking: In the following examples, we should *not* emit a diagnostic, but I think your implementation does produce a false-positive diagnostic:
```js
let a = false;
if (a) {
} else (() => console.log("b"))()  // inserting 'if' would be a syntax error
{ console.log("c"); }
```
```js
let a = false;
let o = {};
if (a) {
} else () => {}  // inserting 'if' would be a syntax error
{ console.log("c"); }
```
```js
let a = false;
let o = someExternalVariable;
if (a) {
} else (o.m())  // body of 'else' has potential side effects; inserting 'if' might be unintended
{ console.log("c"); }
```

File: src/quick-lint-js/parse.h [hunk 3]
```diff
@@ -2515,8 +2515,26 @@ class parser {
...
+      switch (this->peek().type) {
+      default:
+        parse_and_visit_body();
+        break;
+
+      case token_type::left_paren:
+        expression *ast = this->parse_expression(precedence{});
+        this->visit_expression(ast, v, variable_context::rhs);
+
+        if (this->peek().type == token_type::left_curly) {
```
Comment (strager) [inline src/quick-lint-js/parse.h:2530]: I think we also need to check the kind of `ast`. See examples in my 'blocking' comment.

In #266, I wrote "side-effect-free expression". I think this is conservative but probably still helpful. "side-effect-free" would be any combination of:
* `binary_operator` (I think?) (if children are side-effect free)
* `cond`
* `_typeof`
* `_class`
* `template` (if children are side-effect free)
* `array` (if children are side-effect free)
* `object` (if children are side-effect free)
* `conditional` (if children are side-effect free)
* ...

In particular:
* `dot` is disallowed because property access can have arbitrary side effects

A `bool is_side_effect_free(expression*)` function would be a welcome addition!

File: src/quick-lint-js/parse.h [hunk 4]
```diff
@@ -2515,8 +2515,26 @@ class parser {
...
+      source_code_span else_span = this->peek().span();
       this->skip();
-      parse_and_visit_body();
+
+      switch (this->peek().type) {
+      default:
+        parse_and_visit_body();
+        break;
+
+      case token_type::left_paren:
```
Comment (strager) [inline src/quick-lint-js/parse.h:2526]: Hmm, this is simpler than I thought. I like your idea of special-casing expression statements beginning with `(`.

**Review by strager — CHANGES_REQUESTED** [review]

File: src/quick-lint-js/parse.h (hunk 3 above)
Comment (keyehzy) [reply src/quick-lint-js/parse.h:2608]: I'm not sure yet about all the base cases, see body of `bool is_side_effect_free(expression* ast)`.

File: src/quick-lint-js/error.h (hunk 1 above)
Comment (keyehzy) [reply src/quick-lint-js/error.h]: Done.

File: src/quick-lint-js/parse.h (hunk 3 above)
Comment (keyehzy) [reply src/quick-lint-js/parse.h:2608]: As for checking the kind of the expression, right now the only invalidating kinds are arrow functions. I'm trying to come up with other examples where it's not caught by the side-effect-free condition.

File: src/quick-lint-js/parse.h [hunk 5]
```diff
@@ -2460,6 +2460,71 @@ class parser {
     v.visit_exit_with_scope();
   }
 
+  bool is_side_effect_free(expression *ast) {
```
Comment (strager) [inline src/quick-lint-js/parse.h:2463]: Nit: Maybe it'd be better to invert the meaning of this function and call it `has_potential_side_effects`? `!is_side_effect_free(ast)` sounds like a double negative to me.

File: src/quick-lint-js/parse.h [hunk 6]
```diff
@@ -2460,6 +2460,71 @@ class parser {
...
+    switch (ast->kind()) {
+    case expression_kind::_class:
```
Comment (strager) [inline src/quick-lint-js/parse.h:2465]: Blocking: Class expressions might have side effects. For example:
```js
function f() { console.log("yo"); }
let C = class { [f()]() {} };
```

We don't have enough information in the AST to know, so let's be conservative and assume that side effects might happen.

File: src/quick-lint-js/parse.h [hunk 7]
```diff
@@ -2460,6 +2460,71 @@ class parser {
...
+    case expression_kind::_invalid:
+    case expression_kind::_typeof:
```
Comment (strager) [inline src/quick-lint-js/parse.h:2467]: Blocking: `typeof` might have a side effect, depending on its argument. For example:
```js
typeof 42; // no side effect
typeof f(); // side effect
```

File: src/quick-lint-js/parse.h [hunk 8]
```diff
@@ -2460,6 +2460,71 @@ class parser {
...
+    case expression_kind::await:
+    case expression_kind::import:
+    case expression_kind::literal:
+    case expression_kind::new_target:
+    case expression_kind::private_variable:
+    case expression_kind::rw_unary_prefix:
+    case expression_kind::rw_unary_suffix:
+    case expression_kind::spread:
```
Comment (strager) [inline src/quick-lint-js/parse.h:2475]: Blocking: `spread` might have side effects, because a spread would call an iterator's `next` function.

File: src/quick-lint-js/parse.h (hunk 5 above)
Comment (strager) [inline src/quick-lint-js/parse.h:2463]: I think this function could easily be tested. Let me know if you want some help writing tests. (I think the tests would go in test-parse.cpp or test-parse-expression.cpp.) I provided some non-obvious/tricky examples in some comments.

File: test/test-parse-statement.cpp [hunk 9]
```diff
@@ -578,6 +578,69 @@ TEST(test_parse, else_without_if) {
...
+    padded_string code(u8"if (a) { b; } else () => {} { c; }"_sv);
+    parser p(&code, &v);
+    EXPECT_TRUE(p.parse_and_visit_statement(v));
+    EXPECT_THAT(v.errors, IsEmpty());
+  }
+
+  {
+    spy_visitor v;
+    padded_string code(
+        u8"if (a) { b; } else (() => console.log(c))() { d; }"_sv);
```
Comment (strager) [inline test/test-parse-statement.cpp:629]: Blocking: This should report a missing-semicolon error.

Perhaps you mean to write this code?
```js
if (a) { b; } else (() => console.log(c))()
{ d; }
```

File: src/quick-lint-js/parse.h [hunk 10]
```diff
@@ -2515,8 +2580,42 @@ class parser {
...
+        if (this->peek().type == token_type::left_curly) {
+          if (!is_invalidating_if && this->is_side_effect_free(ast)) {
+            this->error_reporter_->report(
+                error_else_with_conditional_missing_if{
+                    .else_token = else_span,
+                });
+          }
+          parse_and_visit_body();
+        }
+        break;
```
Comment (strager) [inline src/quick-lint-js/parse.h:2617]: Blocking: After parsing the expression statement, the language says we need a semicolon. (A semicolon might be inserted by ASI if a newline or `}` is found instead of a semicolon.)

In other words, the following code should produce a syntax error (and quick-lint-js does report a syntax error before your commit):
```js
if (a) { b; } else (() => console.log(c))() { d; }
```

File: test/test-parse-statement.cpp [hunk 11]
```diff
@@ -578,6 +578,69 @@ TEST(test_parse, else_without_if) {
...
+    EXPECT_THAT(
+        v.errors,
+        ElementsAre(ERROR_TYPE_FIELD(
+            error_else_with_conditional_missing_if, else_token,
+            offsets_matcher(&code, strlen(u8"if (a) { b; } "), u8"else"))));
+  }
+
+  {
+    spy_visitor v;
+    padded_string code(u8"if (a) { b; } else () => {} { c; }"_sv);
```
Comment (strager) [inline test/test-parse-statement.cpp:620]: Blocking: This should report a missing-semicolon error.

Perhaps you mean to write this code?
```js
if (a) { b; } else () => {}
{ c; }
```

File: test/test-parse-statement.cpp [hunk 12]
```diff
@@ -578,6 +578,69 @@ TEST(test_parse, else_without_if) {
...
+    EXPECT_TRUE(p.parse_and_visit_statement(v));
+    EXPECT_THAT(v.errors, IsEmpty());
+  }
+
+  {
+    spy_visitor v;
+    padded_string code(u8"if (a) { b; } else (o.m()) { c; }"_sv);
+    parser p(&code, &v);
+    EXPECT_TRUE(p.parse_and_visit_statement(v));
+    EXPECT_THAT(v.errors, IsEmpty());
```
Comment (strager) [inline test/test-parse-statement.cpp:640]: Blocking: This should report a missing-semicolon error.

**Review by strager — CHANGES_REQUESTED** [review]
Hmm, this issue is more complicated than I thought! Lots of edge cases.

File: src/parse.cpp [hunk 13]
```diff
@@ -2065,6 +2065,73 @@ expression* parser::maybe_wrap_erroneous_arrow_function(
...
+    return this->has_potential_side_effects(ast->child_0()) ||
+           this->has_potential_side_effects(ast->child_1()) ||
+           this->has_potential_side_effects(ast->child_2());
+
+  case expression_kind::object: {
+    for (int i = 0; i < ast->object_entry_count(); i++) {
+      auto entry = ast->object_entry(i);
+      if (entry.property.has_value()) {
+        if (this->has_potential_side_effects(*entry.property) ||
+            this->has_potential_side_effects(entry.value))
```
Comment (strager) [inline src/parse.cpp:2125]: Shouldn't we check `entry.value` regardless of `entry.property.has_value()`?

File: test/test-parse-expression.cpp [hunk 14]
```diff
@@ -3161,6 +3161,273 @@ TEST_F(test_parse_expression,
   }
 }
 
+TEST(test_parse, test_expression_for_potential_side_effects) {
+  {
+    test_parser p(u8"class {};"_sv);
+    expression* ast = p.parse_expression();
+    bool side_effects = p.parser().has_potential_side_effects(ast);
+    EXPECT_TRUE(side_effects);
+  }
```
Comment (strager) [inline test/test-parse-expression.cpp:3170]: This really should be `false`, but I know we can't implement that right now. Perhaps the following examples are better, because they do have side effects:
```js
class { [f()]() { } }
```
or:
```js
class { static pi = calculatePi(); }
```

File: test/test-parse-expression.cpp [hunk 15]
```diff
@@ -3161,6 +3161,273 @@ TEST_F(test_parse_expression,
...
+
+  {
+    test_parser p(u8"await foo;"_sv);
+    expression* ast = p.parse_expression();
+    bool side_effects = p.parser().has_potential_side_effects(ast);
+    EXPECT_TRUE(side_effects);
+  }
+
+  {
+    test_parser p(u8"import foo;"_sv);
```
Comment (strager) [inline test/test-parse-expression.cpp:3194]: Blocking: This has syntax errors. I think you meant this:
```js
import(foo);
```

File: test/test-parse-expression.cpp [hunk 16]
```diff
@@ -3161,6 +3161,273 @@ TEST_F(test_parse_expression,
...
+
+  {
+    test_parser p(u8"x++;"_sv);
+    expression* ast = p.parse_expression();
+    bool side_effects = p.parser().has_potential_side_effects(ast);
+    EXPECT_TRUE(side_effects);
+  }
+
+  {
+    test_parser p(u8"...foo"_sv);
```
Comment (strager) [inline test/test-parse-expression.cpp:3215]: Blocking: This is a syntax error. Spread isn't allowed on its own. I think you meant one of the following:
```js
{...foo}
```
```js
[...foo]
```
```js
f(...foo)
```

File: test/test-parse-expression.cpp [hunk 17]
```diff
@@ -3161,6 +3161,273 @@ TEST_F(test_parse_expression,
...
+    test_parser p(u8"new foo;"_sv);
+    expression* ast = p.parse_expression();
+    bool side_effects = p.parser().has_potential_side_effects(ast);
+    EXPECT_TRUE(side_effects);
+  }
+
+  {
+    // TODO(keyehzh): Need more specific tests as some binary operators might
+    // have side-effects (such as '+', '==', ...) and some don't
+    // (such as '===' and '!==')
```
Comment (strager) [inline test/test-parse-expression.cpp:3311]: :+1:

File: test/test-parse-expression.cpp [hunk 18]
```diff
@@ -3161,6 +3161,273 @@ TEST_F(test_parse_expression,
...
+    expression* ast = p.parse_expression();
+    bool side_effects = p.parser().has_potential_side_effects(ast);
+    EXPECT_TRUE(side_effects);
+  }
+
+  {
+    test_parser p(u8"{[f()]: true}"_sv);
+    expression* ast = p.parse_expression();
+    bool side_effects = p.parser().has_potential_side_effects(ast);
+    EXPECT_TRUE(side_effects);
```
Comment (strager) [inline test/test-parse-expression.cpp:3427]: Nit: I think we can save on lines of code:
```suggestion
    EXPECT_TRUE(p.parser().has_potential_side_effects(ast));
```

File: test/test-parse-expression.cpp [hunk 19]
```diff
@@ -3161,6 +3161,273 @@ TEST_F(test_parse_expression,
...
+    expression* ast = p.parse_expression();
+    bool side_effects = p.parser().has_potential_side_effects(ast);
+    EXPECT_FALSE(side_effects);
+  }
+
+  {
+    test_parser p(u8"{key: f()}"_sv);
+    expression* ast = p.parse_expression();
+    bool side_effects = p.parser().has_potential_side_effects(ast);
+    EXPECT_TRUE(side_effects);
```
Comment (strager) [inline test/test-parse-expression.cpp:3420]: Nit: I think we can safe a line of code by making `has_potential_side_effects` `static`:
```suggestion
    expression* ast = this->parse_expression(u8"{key: f()}"_sv);
    bool side_effects = parser::has_potential_side_effects(ast);
    EXPECT_TRUE(side_effects);
```

**Review by strager — CHANGES_REQUESTED** [review]
Blocking: `if (a) { b; } else (c) { d; }` is still an issue (false negative).

Other than that, this commit looks bueno!

File: src/quick-lint-js/parse.h [hunk 20]
```diff
@@ -2460,6 +2460,71 @@ class parser {
     v.visit_exit_with_scope();
   }
 
+  bool is_side_effect_free(expression *ast) {
+    switch (ast->kind()) {
+    case expression_kind::_class:
+    case expression_kind::_invalid:
+    case expression_kind::_typeof:
+    case expression_kind::await:
+    case expression_kind::import:
```
[the remark this answers is not in the document]
Comment (keyehzy) [reply src/quick-lint-js/parse.h]: Fixed.

File: src/quick-lint-js/parse.h [hunk 21]
```diff
@@ -2460,6 +2460,71 @@ class parser {
...
+    case expression_kind::literal:
+    case expression_kind::new_target:
+    case expression_kind::private_variable:
+    case expression_kind::rw_unary_prefix:
+    case expression_kind::rw_unary_suffix:
+    case expression_kind::spread:
+    case expression_kind::super:
+    case expression_kind::unary_operator:
```
[the remark this answers is not in the document]
Comment (keyehzy) [reply src/quick-lint-js/parse.h]: Fixed.

File: src/quick-lint-js/parse.h [hunk 22]
```diff
@@ -2460,6 +2460,71 @@ class parser {
...
+    case expression_kind::variable:
+    case expression_kind::yield_many:
+    case expression_kind::yield_none:
+    case expression_kind::yield_one:
```
[the remark this answers is not in the document]
Comment (keyehzy) [reply src/quick-lint-js/parse.h]: Fixed.

File: src/quick-lint-js/parse.h [hunk 23]
```diff
@@ -2460,6 +2460,71 @@ class parser {
...
+      return true;
+
+    case expression_kind::dot:
+    case expression_kind::index:
+    case expression_kind::function:
+    case expression_kind::named_function:
+      return false;
+
+    case expression_kind::_new:
```
[the remark this answers is not in the document]
Comment (keyehzy) [reply src/quick-lint-js/parse.h]: Fixed.

File: src/parse.cpp [hunk 24]
```diff
@@ -2065,6 +2065,73 @@ expression* parser::maybe_wrap_erroneous_arrow_function(
...
+    return has_potential_side_effects(ast->child_0()) ||
+           has_potential_side_effects(ast->child_1()) ||
+           has_potential_side_effects(ast->child_2());
+
+  case expression_kind::object: {
+    for (int i = 0; i < ast->object_entry_count(); i++) {
+      auto entry = ast->object_entry(i);
+      if (entry.property.has_value()) {
+        if (has_potential_side_effects(*entry.property) ||
+            has_potential_side_effects(entry.value))
```
Comment (strager) [inline src/parse.cpp:2125]: Shouldn't we check `entry.value` regardless of `entry.property.has_value()`?

File: src/parse.cpp [hunk 25]
```diff
@@ -2065,6 +2065,73 @@ expression* parser::maybe_wrap_erroneous_arrow_function(
...
+          return true;
+      }
+    }
+    return false;
+  }
+  }
+  return false;
```
Comment (strager) [inline src/parse.cpp:2132]: Nit:
```suggestion
  QLJS_UNREACHABLE();
```

File: src/quick-lint-js/parse.h [hunk 26]
```diff
@@ -2515,8 +2517,43 @@ class parser {
...
+        default:
+          break;
+
+        case expression_kind::arrow_function_with_expression:
+        case expression_kind::arrow_function_with_statements:
+          is_invalidating_if = true;
+          break;
+        }
+
+        this->consume_semicolon();
```
Comment (strager) [inline src/quick-lint-js/parse.h:2544]: Blocking: This looks like it's in the wrong spot. Won't skipping a semicolon here cause the following code to have the error_else_with_conditional_missing_if warning?
```js
if (a) { b; } else (c);
{ d; }
```

I think we need to consume a semicolon after the check on line 2546.

File: test/test-parse-statement.cpp [hunk 27]
```diff
@@ -578,6 +578,135 @@ TEST(test_parse, else_without_if) {
...
+    EXPECT_THAT(
+        v.errors,
+        ElementsAre(
+            ERROR_TYPE_FIELD(
+                error_missing_semicolon_after_statement, where,
+                offsets_matcher(&code, strlen(u8"if (a) { b; } else (c)"),
+                                u8"")),
+            ERROR_TYPE_FIELD(
+                error_else_with_conditional_missing_if, else_token,
+                offsets_matcher(&code, strlen(u8"if (a) { b; } "), u8"else"))));
```
Comment (strager) [inline test/test-parse-statement.cpp:619]: :+1:

**Review by strager — CHANGES_REQUESTED** [review]

File: test/test-parse-expression.cpp (hunk 14 above)
Comment (keyehzy) [reply test/test-parse-expression.cpp]: This one is resolved.

File: test/test-parse-expression.cpp (hunk 15 above)
Comment (keyehzy) [reply test/test-parse-expression.cpp]: This one is resolved.

File: src/quick-lint-js/parse.h (hunk 26 above)
Comment (keyehzy) [reply src/quick-lint-js/parse.h]: Agreed!

File: src/parse.cpp (hunk 25 above)
Comment (keyehzy) [reply src/parse.cpp]: Thanks!

File: src/parse.cpp [hunk 28]
```diff
@@ -2121,15 +2121,14 @@ bool parser::has_potential_side_effects(expression* ast) {
     for (int i = 0; i < ast->object_entry_count(); i++) {
       auto entry = ast->object_entry(i);
       if (entry.property.has_value()) {
-        if (has_potential_side_effects(*entry.property) ||
-            has_potential_side_effects(entry.value))
-          return true;
+        return has_potential_side_effects(*entry.property) ||
+               has_potential_side_effects(entry.value);
```
Comment (strager) [inline src/parse.cpp:2125]: Blocking: You're only checking the first entry which has a value! We should check all entries.

File: src/quick-lint-js/parse.h (hunk 26 above)
Comment (strager) [reply src/quick-lint-js/parse.h]: Please add a test for this example.

**Comment by strager** [comment]
There are merge conflicts because `parser::parse_and_visit_if`'s body has been moved to `src/quick-lint-js/parse-statement-inl.h`.
